inference-engine/src/mkldnn_plugin/mkldnn_graph_optimizer.cpp

```diff
@@ -2342,3 +2298,81 @@ void MKLDNNGraphOptimizer::MergePermuteAndReorder(MKLDNNGraph &graph) {
         }
     }
 }
+
+static bool isReorderAvailable(const MKLDNNEdgePtr parentEdge, const MKLDNNEdgePtr childEdge, const mkldnn::engine& eng) {
+    MKLDNNMemoryDesc dstMemDesc(childEdge->getDesc());
+    MKLDNNMemoryDesc srcMemDesc(parentEdge->getDesc());
+
+    memory::primitive_desc dstPrimitiveDesc(dstMemDesc, eng);
+    memory::primitive_desc srcPrimitiveDesc(srcMemDesc, eng);
+    mkldnn::primitive_attr attr;
+
+    mkldnn_primitive_desc_t result = nullptr;
+    auto status = mkldnn_reorder_primitive_desc_create_v2(&result, srcPrimitiveDesc.get(), dstPrimitiveDesc.get(), attr.get());
+    if (result) {
+        mkldnn_primitive_desc_destroy(result);
+    }
+
+    return mkldnn_success == status;
+}
+
+void MKLDNNGraphOptimizer::AddConvertToReorder(MKLDNNGraph &graph) {
+    const auto& vecNodes = graph.GetNodes();
+    std::set<MKLDNNNodePtr> insertedReorders;
+
+    for (int i = 0; i < vecNodes.size(); i++) {
+        auto node = vecNodes[i];
+        if (Reorder == node->getType()) {
+            auto inpEdge = node->getParentEdgeAt(0);
+            auto outEdge = node->getChildEdgeAt(0);
+            // Some guard checks
+            if (inpEdge->getDesc().getPrecision() == outEdge->getDesc().getPrecision()) {
+                continue;
+            }
+
+            if (insertedReorders.count(node)) {
+                continue;
+            }
+
+            // Check if there is a reorder that supports the type conversion
+            if (isReorderAvailable(inpEdge, outEdge, graph.getEngine())) {
+                continue;
+            }
+
+            std::string convertName = inpEdge->getParent()->getName() + "_" +
+                    inpEdge->getDesc().getPrecision().name() + "_" + outEdge->getDesc().getPrecision().name();
+            //Insert Convert before the reorder
+            // make CNNLayer
+            CNNLayerPtr convert(new  CNNLayer(LayerParams{convertName, "Convert", outEdge->getDesc().getPrecision()}));
+
+            auto oIndex = outEdge->getOutputNum();
+            auto iIndex = inpEdge->getInputNum();
+            if (iIndex < 0 || oIndex < 0)
```

Comment:
I understand that if we will not throw exception here it will be thrown inside reorder because of unsupported parameters. However for me this totally strange that graph optimization throws exception if it cannot optimize smt. Seems like it is still better to see second exception.
I would suggest to fix `iIndex == 0 && oIndex == 0` limitation in this PR.

Reply:
Yes, it makes sense, but anyways this chunk of code has been removed (see b995de5).